Objects/typeobject.c

```diff
@@ -5288,27 +5288,58 @@ _PyType_GetModuleByDef2(PyTypeObject *left, PyTypeObject *right,
 }
 
 
+// PyType_GetBaseByToken() family are better to be placed from most to least
+// frequent in the src and obj files (the decls in random order) for PGO on
+// MSVC. They'll be optimized not for speed if a cold sub function precedes,
+// in which case the unrelated functions above can be slower if there are
+// common conds such as `if (!_PyType_HasFeature(tp, Py_TPFLAGS_HEAPTYPE))`.
+static int
+check_base_by_token(PyTypeObject *, void *);
+static inline PyTypeObject *
+get_base_by_token_from_mro(PyTypeObject *, void *);
 static PyTypeObject *
-get_base_by_token_recursive(PyTypeObject *type, void *token)
+get_base_by_token_recursive(PyTypeObject *, void *);
+
+int
+PyType_GetBaseByToken(PyTypeObject *type, void *token, PyTypeObject **result)
 {
-    assert(PyType_GetSlot(type, Py_tp_token) != token);
-    PyObject *bases = lookup_tp_bases(type);
-    assert(bases != NULL);
-    Py_ssize_t n = PyTuple_GET_SIZE(bases);
-    for (Py_ssize_t i = 0; i < n; i++) {
-        PyTypeObject *base = _PyType_CAST(PyTuple_GET_ITEM(bases, i));
-        if (!_PyType_HasFeature(base, Py_TPFLAGS_HEAPTYPE)) {
-            continue;
-        }
-        if (((PyHeapTypeObject*)base)->ht_token == token) {
-            return base;
-        }
-        base = get_base_by_token_recursive(base, token);
-        if (base != NULL) {
-            return base;
-        }
+    if (result == NULL) {
+        // If the `result` is checked only once here, the subsequent
+        // branches will become trivial to optimize.
+        return check_base_by_token(type, token);
+    }
+    if (token == NULL || !PyType_Check(type)) {
+        *result = NULL;
+        return check_base_by_token(type, token);
+    }
+
+    if (!_PyType_HasFeature(type, Py_TPFLAGS_HEAPTYPE)) {
+        // No static type has a heaptype superclass,
+        // which is ensured by type_ready_mro().
+        *result = NULL;
```

Comment:
Maybe you could add
```C
L5316-5320
    if (!_PyType_HasFeature(type, Py_TPFLAGS_HEAPTYPE)) {
        // No static type has a heaptype superclass,
        // which is ensured by type_ready_mro().
        *result = NULL;
        return 0;
    }

L5339-5341
    else {
        *result = NULL;
        return 0;
    }
    }
```
To the goto block, kinda like this
```C
if (!_PyType_HasFeature(type, Py_TPFLAGS_HEAPTYPE)) {
      goto error;
}
...
error:
      *result = NULL;
      return 0;
```
It might cut down on some redundant code. I think that's good.